src/symbolize/gimli/lru.rs

```diff
@@ -0,0 +1,73 @@
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// least-recently-used cache with static size
+pub(crate) struct Lru<T, const N: usize> {
+    // SAFETY: len <= initialized values
+    len: usize,
+    arr: [MaybeUninit<T>; N],
+}
+
+impl<T, const N: usize> Default for Lru<T, N> {
+    fn default() -> Self {
+        Lru {
+            len: 0,
+            arr: [const { MaybeUninit::uninit() }; N],
+        }
+    }
+}
+
+impl<T, const N: usize> Lru<T, N> {
+    #[inline]
+    pub fn clear(&mut self) {
+        let len = self.len;
+        self.len = 0;
+        // SAFETY: we can't touch these values again due to setting self.len = 0
+        unsafe { ptr::drop_in_place(ptr::addr_of_mut!(self.arr[0..len]) as *mut [T]) }
+    }
+
+    #[inline]
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        self.arr[0..self.len]
+            .iter()
+            // SAFETY: we only iterate initialized values due to our len invariant
+            .map(|init| unsafe { init.assume_init_ref() })
+    }
+
+    #[inline]
+    pub fn push_front(&mut self, value: T) -> Option<&mut T> {
+        if N == 0 {
+            return None;
+        } else if self.len == N {
+            self.len = N - 1;
+            // SAFETY: we maintain len invariant and bail on N == 0
+            unsafe { ptr::drop_in_place(self.arr.as_mut_ptr().cast::<T>().add(N - 1)) };
+        };
+        let len_to_init = self.len + 1;
+        let mut last = MaybeUninit::new(value);
+        for elem in self.arr[0..len_to_init].iter_mut() {
+            last = mem::replace(elem, last);
```

Comment:
Isn't this just `mem::swap`?

Reply:
what we do is assign the `last` to an index, and then take out the old value. it isn't really `mem::swap` because we're instead moving everything a step over, and that isn't really expressible that way (and there's no `windows_mut`, besides).

Reply:
Maybe I'm having a brain fart but I mean literally `last = mem::replace(elem, last);` looks suspiciously like `mem::swap(elem, &mut last);` no?

Reply:
oh.

Reply:
yes.

Reply:
Hm, that had a big effect on code size???